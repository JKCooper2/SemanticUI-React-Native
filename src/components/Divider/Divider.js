// @flow
import React, { PureComponent } from 'react';
import type { Node } from 'react';
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import { View, Text, StyleSheet } from 'react-native';
import _ from 'lodash';

import { spacing, fonts, colors } from '~/styles';

type Props = {
    children?: Node,
    style?: StyleObj,
    fitted?: boolean,
    hidden?: boolean,
    section?: boolean,
    clearing?: boolean,
    inverted?: boolean,
    content?: Node
};

const getHeight = (fitted: boolean, section: boolean, clearing: boolean): StyleObj => {
    if (fitted) {
        return { viewHeight: spacing.padding.tiny, barHeight: spacing.padding.tiny / 2 };
    }

    if (section) {
        return { viewHeight: spacing.padding.large, barHeight: spacing.padding.large / 2 };
    }

    if (clearing) {
        return { viewHeight: spacing.padding.large, barHeight: 0 };
    }

    return { viewHeight: spacing.padding.medium, barHeight: spacing.padding.medium / 2 };
};

const getBarColor = (hidden, inverted) => {
    if (hidden) {
        return colors.colors.transparent;
    }

    if (inverted) {
        return colors.colors.white;
    }

    return colors.colors.lightGrey;
};

class Divider extends PureComponent<Props> {
    static defaultProps = {
        style: {},
        fitted: false,
        hidden: false,
        section: false,
        clearing: false,
        inverted: false,
        children: null,
        content: null
    };

    render() {
        const { children, fitted, hidden, section, clearing, inverted, content } = this.props;

        let { viewHeight, barHeight } = getHeight(fitted, section, clearing);
        let barColor = getBarColor(hidden, inverted);

        let usedChildren = content || children;

        let newChildren = _.isString(usedChildren) ? (
            <Text style={{ paddingHorizontal: spacing.padding.medium, fontSize: fonts.size.default, fontWeight: 'bold', color: barColor }}>{children}</Text>
        ) : children;

        return newChildren ? (
            <View style={{ height: viewHeight, alignItems: 'center', flexDirection: 'row' }}>
                <View style={[{ height: barHeight, borderColor: barColor }, styles.line]} />
                {newChildren}
                <View style={[{ height: barHeight, borderColor: barColor }, styles.line]} />
            </View>
        ) : (
            <View style={{ height: viewHeight, flexDirection: 'row' }}>
                <View style={[{ height: barHeight, borderColor: barColor }, styles.line]} />
            </View>
        );
    }
}

const styles = StyleSheet.create({
    line: {
        borderBottomWidth: StyleSheet.hairlineWidth,
        flex: 1,
        borderStyle: 'solid'
    }
});

export default Divider;