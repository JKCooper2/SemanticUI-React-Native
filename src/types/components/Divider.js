// @flow
import type { Node } from 'react';
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

export type DividerPropsType = {
    children?: Node,
    style?: StyleObj,
    fitted: boolean,
    hidden: boolean,
    section: boolean,
    clearing: boolean,
    inverted: boolean,
    content?: string
};
