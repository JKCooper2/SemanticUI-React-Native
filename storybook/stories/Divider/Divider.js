// @flow
import React, { Fragment } from 'react';
import { Text, View } from 'react-native';

import { storiesOf } from '@storybook/react-native';

import Divider from '~/components/Divider';

const dividerStories = storiesOf('Divider', module);

dividerStories.add('Standard', () => (
    <Fragment>
        <Text>Header Above</Text>
        <Divider />
        <Text>Header Below</Text>
    </Fragment>
));

dividerStories.add('String Child', () => <Divider>Test</Divider>);

dividerStories.add('Content', () => <Divider content="Test" />);

dividerStories.add('React Child', () => (
    <Divider>
        <Text>Test</Text>
    </Divider>
));

dividerStories.add('Fitted', () => (
    <Fragment>
        <Text>Header Above</Text>
        <Divider fitted />
        <Text>Header Below</Text>
    </Fragment>
));

dividerStories.add('Section', () => (
    <Fragment>
        <Text>Header Above</Text>
        <Divider section />
        <Text>Header Below</Text>
    </Fragment>
));

dividerStories.add('Clearing', () => (
    <Fragment>
        <Text>Header Above</Text>
        <Divider clearing />
        <Text>Header Below</Text>
    </Fragment>
));

dividerStories.add('Hidden', () => (
    <Fragment>
        <Text>Header Above</Text>
        <Divider hidden />
        <Text>Header Below</Text>
    </Fragment>
));

dividerStories.add('Inverted', () => (
    <View style={{ backgroundColor: 'black' }}>
        <Text>Header Above</Text>
        <Divider inverted>Inverted</Divider>
        <Text>Header Below</Text>
    </View>
));

export default dividerStories;
