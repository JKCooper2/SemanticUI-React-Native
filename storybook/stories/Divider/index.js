// @flow
import Divider from './Divider';

export {
    Divider
};