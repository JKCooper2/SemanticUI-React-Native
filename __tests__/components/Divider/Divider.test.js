// @flow
import React from 'react';
import { shallow } from 'enzyme';

import Divider from '~/components/Divider';

test('renders correctly', () => {
    const divider = shallow(<Divider />);
    expect(divider).toMatchSnapshot();
});
