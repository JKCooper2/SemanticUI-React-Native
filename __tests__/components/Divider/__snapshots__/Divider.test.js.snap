// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Component
  style={
    Object {
      "flexDirection": "row",
      "height": 20,
    }
  }
>
  <Component
    style={
      Array [
        Object {
          "borderColor": "#d3d3d3",
          "height": 10,
        },
        Object {
          "borderBottomWidth": 0.5,
          "borderStyle": "solid",
          "flex": 1,
        },
      ]
    }
  />
</Component>
`;
